### DISCUSSION TITLE
DataTable: Row stops rendering when Command in column has `update="@id"`

### DISCUSSION THREAD

**Discussion by melloware** [primefaces/community#2350] [PrimeFaces] [answered]
@tandraschko @Rapster thought this would be easier to put in a discussion.    first I am trying everything to reproduce this issue but so far i have not been able to combine the magic but have narrowed down what is happening.

> [!NOTE]  
> This code works fine in 13.0.10 and does not in 14.0.2 with no other changes to the server.

### Environment
Jboss EAP 7.4.12, Java 11, Mojarra 2.3.14

### Scenario

A datatable that has a CommandButton or link with `update="@root:@id(growl)"` causes the row `var` to go NULL and the rest of the columns stop rendering.  NO ERROR in any logs.  Example:

```xml
<p:commandButton value="ClickMe" update="@root:@id(growl)" />
```

The screen renders like this:
![pf-dt-before](https://github.com/user-attachments/assets/9a7ff9d9-496d-46a2-889a-f16b107bf2b6)

Any column after the link is empty now and in debugging the `var="row"` of the Datatable is verified to be NULL like somehow it gets lost.

### Solution I Can't Explain

Changing the `update=""` to not use `@root:@id` and use a plain expression like this...

```xml
<p:commandButton value="ClickMe" update=":frmEntity:growl" />
```

This fixes it and now the table renders properly.

![pf-dt-after](https://github.com/user-attachments/assets/73c46bd5-7bc1-49e7-a2ee-36ecf5ae4b3d)

### Conclusion

The only thing I know is its related to the `update="@root:@id()"` but what I can't explain is...

1. I can't reproduce it in a PF Test
2. It seems related to SearchExpressions since 13.0.10 uses PF custom expressions and 14.0.0 delegates to Mojarra/MyFaces?
3. ???

I was really hoping if someone did a thought excersize why this blows up with no error but stops rendering the datatable and `row` var is definitely NULL.

**Comment by tandraschko** [comment]
`@id` uses visitTree, maybe something isnt restored correctly in DataTable/UIData#visit, dont know.

**Comment by Rapster** [comment]
@melloware Can you attach a reproducer even though you can't produce ?

**Comment by melloware** [comment]
@Rapster here is my shot at reproducer: 
[discussion-2350.zip](https://github.com/user-attachments/files/16311460/discussion-2350.zip)

I have run it with Mojarra 2.3.9-2.3.21 to rule out Mojarra and I did the same with my Jboss app I upgraded from Mojarra 2.3.14 to latest 2.3.21 with no difference.  That is what makes me think it might be a PF subtle bug.

@tandraschko that is a great hint I didn't realize one used `visitTree` and the other flavor doesn't.

**Reply by Rapster** [reply]
Did you try a put ~condition breakpoint on any Exception~ a Java Exception breakpoint? AFAIR JSF impl have a bad habit to swallow exceptions in some cases

**Reply by melloware** [reply]
I couldn't figure out where to put one but I definitely feel like something is throwing an error but it's being captured and no errors in log or anything. Just the row stops rendering and the variable is null.

**Reply by Rapster** [reply]
Sorry, I should have said a Java Exception breakpoint, I don't know about VSCode but on Intellij/Eclipse, you can ensure whenever an exception is thrown it'll catch it

**Reply by melloware** [reply]
I will give that a try on Monday in the code that I can reproduce it.

**Reply by melloware** [reply]
OK I enabled "Break on uncaught exceptions" and it did not break on anything so its not an exception that is uncaught.

If I put it on "Any Exception" I get a whole lot of noise about exceptions that looks like they are being handled and look like expected exceptions but its hard to tell its lots of noise from classloader.

**Reply by Rapster** [reply]
Besides an an exception, it's hard to tell... Another way is to check when setRowIndex is set to -1 and see if that's expected...

**Reply by melloware** [reply]
> Besides an an exception, it's hard to tell... Another way is to check when setRowIndex is set to -1 and see if that's expected...

OK I did some more debugging.  I put a breakpoint on `getUpdate()` of the CommandLink with the `@root` on it and here is where it is in the stack.
![image](https://github.com/user-attachments/assets/ecf4b840-f9d0-471b-8103-5e31e4f5d4ee)

What is interesting the AjaxRequestBuilder builds the exact same update whether its `@root:@id` or `:frm:panel".

It evaluates to the following for both

```js
u:"frmControls:pnlEntityTable"
```

So now it has me thinking its somewhere in the Mojarra code for SearchExpressions. We call this in our PF code.

```java
private AjaxRequestBuilder addExpressions(UIComponent component, String expressions, String key, Set<SearchExpressionHint> hints,
            boolean skipUnredered) {
        if (LangUtils.isNotBlank(expressions)) {
            String resolvedExpressions = SearchExpressionUtils.resolveClientIdsAsString(context, component, expressions, hints,
                    skipUnredered ? ComponentUtils.VISIT_HINTS_SKIP_UNRENDERED : null);
            buffer.append(",").append(key).append(":\"").append(resolvedExpressions).append("\"");
        }

        return this;
    }
```

Which delegates to Mojarra in this line:

```java
 List<String> clientIds = context.getApplication().getSearchExpressionHandler().resolveClientIds(
                SearchExpressionContext.createSearchExpressionContext(context, component, hints, visitHints),
                expression);
```

Let me see if I can debug into that.

**Reply by melloware** [reply]
OK so I think it worked in PF13- is because of our custom IdResolver for `@root:@id` did not do VisitTree it walked the components: 

https://github.com/primefaces/primefaces/blob/348f3e09758766b9d00d850985563293795c2461/primefaces/src/main/java/org/primefaces/expression/impl/IdExpressionResolver.java#L45

**Reply by tandraschko** [reply]
Thats what i said already
Either our UIData is broken but i cant help without reproducer

**Reply by melloware** [reply, marked as answer]
I totally understand i can't reproduce it anywhere but in my production app.  VERY FRUSTRATING!   But I was able to fix it for my client by adding this to their code and now everything goes back to working.  @tandraschko not sure if this is worth putting in to PF core or not since I can't reproduce it.

```java
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import javax.faces.FacesException;
import javax.faces.component.UIComponent;
import javax.faces.component.search.SearchExpressionContext;
import javax.faces.component.search.SearchKeywordContext;
import javax.faces.component.search.SearchKeywordResolver;

import org.primefaces.util.ComponentTraversalUtils;

public class IdExpressionResolver extends SearchKeywordResolver {

   private static final Pattern PATTERN = Pattern.compile("id\\(([\\w-]+)\\)");

   @Override
   public void resolve(final SearchKeywordContext searchKeywordContext, final UIComponent current,
            final String keyword) {
      searchKeywordContext.invokeContextCallback(ComponentTraversalUtils.firstWithId(extractId(keyword), current));
   }

   @Override
   public boolean isResolverForKeyword(final SearchExpressionContext searchExpressionContext, final String keyword) {
      return keyword != null && keyword.startsWith("id(");
   }

   protected String extractId(final String expression) {
      try {
         final Matcher matcher = PATTERN.matcher(expression);

         if (matcher.matches()) {
            return matcher.group(1);
         } else {
            throw new FacesException(
                     "Expression does not match following pattern @id(id). Expression: \"" + expression + "\"");
         }
      } catch (final Exception e) {
         throw new FacesException(
                  "Expression does not match following pattern @id(id). Expression: \"" + expression + "\"", e);
      }
   }
}
```

**Reply by tandraschko** [reply]
No its just a hack around a not reproduceable bug 😅

**Reply by melloware** [reply]
10000% hack!

**Reply by melloware** [reply]
@tandraschko @Rapster i got to the bottom of it but can't explain why it doesn't ALWAYS happen but i think its a bug in Mojarra/MyFaces.

The bug is right here: https://github.com/eclipse-ee4j/mojarra/blob/2a0e8fce6edb5d5c0e74771ddf249f152ec84f8a/impl/src/main/java/com/sun/faces/component/search/SearchKeywordResolverImplId.java#L42-L47

The javadoc even says...

>  // Avoid visit tree because in this case we need real component instances.
> // This means components inside UIData will not be scanned.

However I have grepped both MyFaces and Mojarra and cannot find a single instance of where `SearchExpressionHint.SKIP_VIRTUAL_COMPONENTS` is ever used to tell it that this is a UIData?

Stepping through the debugger i verify it never hits that first block and it always uses `visitTree`...

**Comment by wlhUser** [comment]
@melloware I'm currently also struggling with this problem, and try to find workarounds. If you are interested in a simple reproducer, I might have one.

**Reply by melloware** [reply]
yes would LOVE one!  I tried forever to reproduce it and gave up and just implemented the above fix in my production code!

**Reply by wlhUser** [reply]
Here is a reproducer. I hope it works in your environment as well. 
It is based on the PF 15.0.5 showcase, which I ran with the profile **-Pnon-ee,mojarra-4.0**

In the showcase you need to replace the file:
_primefaces-showcase/src/main/webapp/ui/data/datatable/basic.xhtml_
by the content of the attached zip file.
[basic.xhtml.zip](https://github.com/user-attachments/files/21022104/basic.xhtml.zip)

When opening the page of the showcase DataTable -> Basic you see a data table where only the 1st columns contains something, as the others are broken, due to the @id problem.

In the xhtml file you find in the 1st col of the table different versions of the command-link, which are commented.
In the comments you see which ones are working and which not.
It is quite interesting that only the ones which use the PF 'update' attribute fail. The ones using the jsf 'render' work. 
The commandLinks themselves do nothing, to keep the example simple.
[
![screenshot](https://github.com/user-attachments/assets/fccb8c84-6be7-49a2-948a-306ef7ac8634)
](url)

*Notice*: The <p:dialog/> with the embedded datatable at the bottom was necessary in my tests to get the bug. without it it does not occur.

Hope this example helps finding the cause of the problem.

**Reply by melloware** [reply]
This is excellent let me try and debug.

**Reply by melloware** [reply]
Thanks put together a detailed report: https://github.com/primefaces/primefaces/issues/13967 and reproducer
